# Homework

### Question 4

Download the data (only for podcasts). How many records are there?

In [1]:
import homework as docs

github_data = docs.read_github_data('DataTalksClub', 'datatalksclub.github.io')
parsed_data = docs.parse_data(github_data)


In [2]:
len(github_data)

184

A/ `184 records`

### Question 5

Let's prepare this data. It's already structured, so you can chunk it using paragraphs. Let's do chunk size 30 and overlap 15. How many chunks do you have in the result?

In [152]:
import re

paragraphs = []

for data in github_data:
    text = data.content
    paragraphs.extend(re.split(r"\n\s*\n", text.strip()))

In [153]:
len(paragraphs)

854

In [154]:
paragraphs[0]

"---\nepisode: 8\nguests:\n- jekaterinakokatjuhha\nids:\n  anchor: The-Journey-of-a-Data-Generalist-From-Bioinformatics-to-Freelancing---Jekaterina-Kokatjuhha-e1upvim\n  youtube: FRi0SUtxdMw\nimage: images/podcast/s12e08-journey-of-data-generalist-from-bioinformatics-to-freelancing.jpg\nlinks:\n  anchor: https://anchor.fm/datatalksclub/episodes/The-Journey-of-a-Data-Generalist-From-Bioinformatics-to-Freelancing---Jekaterina-Kokatjuhha-e1upvim\n  apple: https://podcasts.apple.com/us/podcast/the-journey-of-a-data-generalist-from/id1541710331?i=1000599125044\n  spotify: https://open.spotify.com/episode/5fB185hGlGYQmdk0kbIsPv?si=YtnsaYNzTc-fl7emZ2IjEA\n  youtube: https://www.youtube.com/watch?v=FRi0SUtxdMw\nseason: 12\nshort: 'The Journey of a Data Generalist: From Bioinformatics to Freelancing'\ntitle: 'The Journey of a Data Generalist: From Bioinformatics to Freelancing'\ntranscript:\n- line: This week we'll talk about being a data generalist. We'll discuss going from\n    bioinformatics

In [156]:
def sliding_window(seq, size, step):
    """Create overlapping chunks using sliding window approach."""
    if size <= 0 or step <= 0:
        raise ValueError("size and step must be positive")

    n = len(seq)
    result = []
    for i in range(0, n, step):
        batch = seq[i:i+size]
        result.append(batch)
        if i + size >= n:
            break

    return result

def format_chunk(chunk):
    return {
        'content': ' '.join(chunk)
    }

In [157]:
chunks = []

for chunk in sliding_window(paragraphs, 30, 15):
    processed = format_chunk(chunk)
    chunks.append(processed)


In [158]:
len(chunks)

56

A/ `56 chunks`

### Question 6

Index the data with Index from minsearch. What's the first episode in the results for "how do I make money with AI?"

In [159]:
chunks[0]

{'content': '---\nepisode: 8\nguests:\n- jekaterinakokatjuhha\nids:\n  anchor: The-Journey-of-a-Data-Generalist-From-Bioinformatics-to-Freelancing---Jekaterina-Kokatjuhha-e1upvim\n  youtube: FRi0SUtxdMw\nimage: images/podcast/s12e08-journey-of-data-generalist-from-bioinformatics-to-freelancing.jpg\nlinks:\n  anchor: https://anchor.fm/datatalksclub/episodes/The-Journey-of-a-Data-Generalist-From-Bioinformatics-to-Freelancing---Jekaterina-Kokatjuhha-e1upvim\n  apple: https://podcasts.apple.com/us/podcast/the-journey-of-a-data-generalist-from/id1541710331?i=1000599125044\n  spotify: https://open.spotify.com/episode/5fB185hGlGYQmdk0kbIsPv?si=YtnsaYNzTc-fl7emZ2IjEA\n  youtube: https://www.youtube.com/watch?v=FRi0SUtxdMw\nseason: 12\nshort: \'The Journey of a Data Generalist: From Bioinformatics to Freelancing\'\ntitle: \'The Journey of a Data Generalist: From Bioinformatics to Freelancing\'\ntranscript:\n- line: This week we\'ll talk about being a data generalist. We\'ll discuss going from\n

In [160]:
from minsearch import Index

index = Index(
    text_fields=["content"],
)

index.fit(chunks)

In [166]:
search = index.search("how do I make money with AI?", num_results=5)

In [182]:
print(search[0])

{'content': '---\nepisode: 8\nguests:\n- jekaterinakokatjuhha\nids:\n  anchor: The-Journey-of-a-Data-Generalist-From-Bioinformatics-to-Freelancing---Jekaterina-Kokatjuhha-e1upvim\n  youtube: FRi0SUtxdMw\nimage: images/podcast/s12e08-journey-of-data-generalist-from-bioinformatics-to-freelancing.jpg\nlinks:\n  anchor: https://anchor.fm/datatalksclub/episodes/The-Journey-of-a-Data-Generalist-From-Bioinformatics-to-Freelancing---Jekaterina-Kokatjuhha-e1upvim\n  apple: https://podcasts.apple.com/us/podcast/the-journey-of-a-data-generalist-from/id1541710331?i=1000599125044\n  spotify: https://open.spotify.com/episode/5fB185hGlGYQmdk0kbIsPv?si=YtnsaYNzTc-fl7emZ2IjEA\n  youtube: https://www.youtube.com/watch?v=FRi0SUtxdMw\nseason: 12\nshort: \'The Journey of a Data Generalist: From Bioinformatics to Freelancing\'\ntitle: \'The Journey of a Data Generalist: From Bioinformatics to Freelancing\'\ntranscript:\n- line: This week we\'ll talk about being a data generalist. We\'ll discuss going from\n

A/ `Season 12 - Episode 8 - The Journey of a Data Generalist From Bioinformatics to Freelancing`

### Question 5 Again

In [218]:
from homework import RawRepositoryFile
from typing import List, Dict, Any
import frontmatter

def parse_data(data_raw: List[RawRepositoryFile]) -> List[Dict[str, Any]]:
    data_parsed = []
    for f in data_raw:
        post = frontmatter.loads(f.content)
        data = post.to_dict()
        data['filename'] = f.filename
        data['content'] = data.get('content', '') + ''.join([t.get('line', '') for t in data.get('transcript', [])])
        data_parsed.append(data)

    return data_parsed

In [219]:
parsed_data_2 = parse_data(github_data)

In [221]:
# [{'content': p } for p in paragraphs]
chunks2 = docs.chunk_documents(parsed_data_2, 30, 15)

In [222]:
len(chunks2)

518519

A/ `518519 chunks` 

### Question 6 Again

In [223]:
index2 = Index(
    text_fields=["content"],
)

index2.fit(chunks2)

In [229]:
index2.search("how do I make money with AI?", num_results=5)[0]

{'start': 28410,
 'content': 'ch money do they make with the',
 'title': 'The Essentials of Public Speaking for Career in Data Science',
 'short': 'The Essentials of Public Speaking for Career in Data Science',
 'guests': ['bentaylor'],
 'image': 'images/podcast/s02e10-public-speaking.jpg',
 'season': 2,
 'episode': 10,
 'ids': {'youtube': 'wOFvlR9UBxI',
  'anchor': 'The-Essentials-of-Public-Speaking-for-Career-in-Data-Science---Ben-Taylor-et0m4p'},
 'links': {'youtube': 'https://www.youtube.com/watch?v=wOFvlR9UBxI',
  'anchor': 'https://anchor.fm/datatalksclub/episodes/The-Essentials-of-Public-Speaking-for-Career-in-Data-Science---Ben-Taylor-et0m4p',
  'spotify': 'https://open.spotify.com/episode/4QWfObiuYmtOCtpSL5LZf9',
  'apple': 'https://podcasts.apple.com/us/podcast/essentials-public-speaking-for-career-in-data-science/id1541710331?i=1000513669829'},
 'transcript': [{'line': "This week, we'll talk about public speaking. We have a special guest today, Ben Taylor. Ben is the chief A

In [230]:
[p['title'] for p in index2.search("how do I make money with AI?", num_results=5)]

['The Essentials of Public Speaking for Career in Data Science',
 'Responsible and Explainable AI',
 'Freelancing in Machine Learning',
 'From Open-Source Maintainer to Founder',
 'Lessons Learned from Freelancing and Working in a Start-up']

A/ `The Essentials of Public Speaking for Career in Data Science`